### PR TITLE
[#3688] Update Objects API registration options form

### PR DESCRIPTION
**Pull request by Viicos**
Part of #3688.

To be resolved:
- https://github.com/open-formulieren/open-forms/pull/3847#discussion_r1483191636
- https://github.com/open-formulieren/open-forms/pull/3847#discussion_r1483222824
  I'm having an issue right now: as the selects are now required, a default objecttype gets selected as soon as the page loads. However, it does not trigger the `useAsync` for versions, resulting in an empty dropdown for versions.

### CODE REVIEW

**Comment by Viicos** [comment]
Of course, https://github.com/open-formulieren/open-forms/pull/3861/commits/f74bc8e492d2091fabd516ae799ed3c5182bd3ed was too easy..
[Screencast from 16-02-24 15:15:50.webm](https://github.com/open-formulieren/open-forms/assets/65306057/82012a04-eecd-499d-9701-3f294a3bb203)

File: src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectsApiOptionsFormFields.js [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+
+  const [availableObjecttypes, setAvailableObjecttypes] = useState([]);
+  const [availableObjecttypeVersions, setAvailableObjecttypeVersions] = useState([]);
+
+  const v2SwitchMessage = intl.formatMessage({
+    defaultMessage: `Switching to the new registration options will remove the existing JSON templates.
+    You will also not be able to save the form until the variables are correctly mapped.
+    Are you sure you want to continue?
+    `,
+    description: 'V2 Switch warning message',
```
Comment (sergei-maertens) [inline src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectsApiOptionsFormFields.js:48]: this needs more context - this is the sole description that will be present in the message catalog.

```suggestion
    description: 'Objects API registration backend: v2 switch warning message',
```

**Comment by Viicos** [comment]
I added test but they are probably meaningless for now as SB is getting crazy when running them I have no idea why

File: src/openforms/js/components/admin/form_design/registrations/index.js [hunk 2]
```diff
@@ -14,20 +15,8 @@ export const BACKEND_OPTIONS_FORMS = {
...
-      },
-      paymentStatusUpdateJson: {
-        'ui:widget': 'textarea',
-        'ui:options': {
-          rows: 5,
-        },
-      },
-    },
+    form: ObjectsApiOptionsForm,
+    onStepEdit: onStepEdit,
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/openforms/js/components/admin/form_design/registrations/index.js:19]: I confirmed the issue and patched straight to master: fac0afc091124bc99d8aa918eccb3c8d11a3f4c8

File: src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectsApiOptionsFormFields.stories.js [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+export const SwitchToV2Empty = {
+  play: async ({canvasElement}) => {
+    window.confirm = jest.fn(() => true);
+    const canvas = within(canvasElement);
+
+    const v2Tab = canvas.getByRole('tab', {selected: false});
+    userEvent.click(v2Tab);
+
+    await waitFor(async () => {
```
Comment (sergei-maertens) [inline src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectsApiOptionsFormFields.stories.js:80]: does everything really need to be wrapped in `waitFor`? I think I'd do a `findByRole` to wait until all options are present and then make assertions on the values:

```js
await canvas.findByRole('option', {name: 'Tree'});
expect(canvas.getByLabelText('Objecttype')).toHaveValue('https://...');
```

The `findByRole` encapsulates the async bit/loading from API endpoint, while asserting functionaly that available options are displayed & eventually the first option is selected.

File: src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectsApiOptionsFormFields.stories.js [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    });
+
+    await waitFor(async () => {
+      const objecttypeVersionSelect = canvas.getByLabelText('Objecttype version');
+
+      await expect(objecttypeVersionSelect.value).toBe('2');
+    });
+
+    const v1Tab = canvas.getByRole('tab', {selected: false});
+    userEvent.click(v1Tab);
```
Comment (sergei-maertens) [inline src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectsApiOptionsFormFields.stories.js:95]:
```suggestion
    await userEvent.click(v1Tab);
```
backend is on a recent storybook version where testing-library/user-events is everything async

**Review by sergei-maertens — CHANGES_REQUESTED** [review]

File: src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectsApiOptionsFormFields.stories.js (hunk 3 above)
Comment (Viicos) [reply src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectsApiOptionsFormFields.stories.js:80]: I always forgot how the API works..

Comment (Viicos) [reply src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectsApiOptionsFormFields.stories.js:80]: Anyway tests are flaky even with this, seems to be a weird race condition again (how surprising)

File: src/openforms/js/components/admin/form_design/registrations/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Viicos) [reply src/openforms/js/components/admin/form_design/registrations/index.js:19]: Luckily the objects API does not require any changes on step edit?

[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/openforms/js/components/admin/form_design/registrations/index.js]: mapping of variable -> json path will need to be updated accordingly when a form variable is deleted/modified ;)

File: src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectsApiOptionsFormFields.stories.js (hunk 3 above)
Comment (sergei-maertens) [reply src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectsApiOptionsFormFields.stories.js]: the lack of `await` before `userEvent.click` may have something to do with that :)

**Review by sergei-maertens — APPROVED** [review]

**Comment by Viicos** [comment]
Wait what how are tests passing

File: src/openforms/js/compiled-lang/en.json [hunk 5]
```diff
@@ -3121,12 +3175,6 @@
       "value": "Products Component"
     }
   ],
-  "aWwaAq": [
-    {
-      "type": 0,
-      "value": "Maximum selected checkboxes"
-    }
-  ],
```
Comment (Viicos) [inline src/openforms/js/compiled-lang/en.json:3129]: Not sure why these where removed

File: src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectTypeSelect.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+};
+
+ObjectTypeSelect.propTypes = {
+  /**
+   * URL to the objecttype in the Object types API.
+   */
+  objecttype: PropTypes.string.isRequired,
+  onChange: PropTypes.func.isRequired,
+  onLoadingError: PropTypes.func.isRequired,
+};
```
Comment (sergei-maertens) [inline src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectTypeSelect.js:58]: this doesn't properly describe the actual props used

File: src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectTypeSelect.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
+import PropTypes from 'prop-types';
+import React, {useEffect} from 'react';
+
+import Select from 'components/admin/forms/Select';
+
+// TODO: properly translate this
+const LOADING_OPTION = [['...', 'loading...']];
```
Comment (sergei-maertens) [inline src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectTypeSelect.js:7]: TODO: :)

File: src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectTypeVersionSelect.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
+import React, {useEffect} from 'react';
+import useAsync from 'react-use/esm/useAsync';
+
+import Select from 'components/admin/forms/Select';
+import {get} from 'utils/fetch';
+
+// TODO: properly translate this
+const LOADING_OPTION = [['...', 'loading...']];
```
Comment (sergei-maertens) [inline src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectTypeVersionSelect.js:8]: TODO

File: src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectTypeVersionSelect.js [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        name="objecttypeVersion"
+        choices={choices}
+        value={selectedVersion ?? ''}
+        onChange={onChange}
+      />
+    </>
+  );
+};
+
+export default ObjectTypeVersionSelect;
```
Comment (sergei-maertens) [inline src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectTypeVersionSelect.js:82]: proptypes missing

File: src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectTypeVersionSelect.js [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  return (
+    <>
+      <Select
+        id="root_objecttypeVersion"
+        name="objecttypeVersion"
+        choices={choices}
+        value={selectedVersion ?? ''}
+        onChange={onChange}
+      />
+    </>
```
Comment (sergei-maertens) [inline src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectTypeVersionSelect.js:78]:
```suggestion
    <Select
      id="root_objecttypeVersion"
      name="objecttypeVersion"
      choices={choices}
      value={selectedVersion ?? ''}
      onChange={onChange}
    />
```

File: src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectsApiOptionsFormFields.js [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+              onChange={onFieldChange}
+            />
+          </Wrapper>
+        </TabPanel>
+      </Tabs>
+    </>
+  );
+};
+
+export default ObjectsApiOptionsFormFields;
```
Comment (sergei-maertens) [inline src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectsApiOptionsFormFields.js:102]: proptypes missing

File: src/openforms/js/components/admin/form_design/registrations/objectsapi/V2ConfigFields.js [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+          onChange={onChange}
+        />
+      </CustomFieldTemplate>
+    </>
+  );
+};
+
+V2ConfigFields.propTypes = {
+  index: PropTypes.number.isRequired,
+};
```
Comment (sergei-maertens) [inline src/openforms/js/components/admin/form_design/registrations/objectsapi/V2ConfigFields.js:248]: incomplete prop types :)

**Review by sergei-maertens — CHANGES_REQUESTED** [review]
Some polishing and then this can be merged.

File: src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectTypeSelect.js (hunk 7 above)
Comment (Viicos) [reply src/openforms/js/components/admin/form_design/registrations/objectsapi/ObjectTypeSelect.js:7]: I will resolve them in another PR as I'm using this loading option elsewhere

**Review by sergei-maertens — APPROVED** [review]

**Pull request merged**